File: 198.py
// Time Complexity : O(n) 
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


class Solution:
    def rob(self, nums: List[int]) -> int:
        arr = []
        if nums == []:
            return 0
        arr.append(0)
        arr.append(nums[0])
        for i in range(2,len(nums)+1):
            arr.append( max(arr[i-1],arr[i-2]+nums[i-1]))
        return arr[-1]
                

File: 198.java
//o(n)
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        } else if (nums.length == 2) {
            return Math.max(nums[0], nums[1]);
        }
        
        int[] tab = new int[nums.length];
        tab[0] = nums[0];
        tab[1] = Math.max(nums[0], nums[1]);
        
        int max = Math.max(tab[0], tab[1]);
        for (int i = 2; i < nums.length; i++) {
            tab[i] = Math.max(nums[i] + tab[i-2], tab[i-1]);
            max = (max < tab[i])? tab[i] : max;
        }
        
        return max;
    }
}